task_planning.py

```diff
@@ -0,0 +1,180 @@
+    )
+
+
+Query.update_forward_refs()
+QueryPlan.update_forward_refs()
+
+
+def query_planner(question: str) -> QueryPlan:
+    messages = [
+        {
+            "role": "system",
+            "content": "You are a world class query planning algorithm capable of breaking apart queries into dependant subqueries, such that the answers can be used to enable the system completing the main query. Do not complete the user query, simply provide a correct compute graph with good specific queries to ask and relevant subqueries. Before completing the list of queries, think step by step to get a better understanding the problem.",
+        },
+        {
+            "role": "user",
+            "content": f"{question}",
+        },
+    ]
+
+    completion = openai.ChatCompletion.create(
+        model="gpt-4-0613",
+        temperature=0,
+        functions=[QueryPlan.openai_schema],
+        function_call={"name": QueryPlan.openai_schema["name"]},
+        messages=messages,
+        max_tokens=1000,
+    )
+    root = QueryPlan.from_response(completion)
+
+    return root
+
+
+def get_execution_order(query_plan: QueryPlan) -> List[int]:
+    """
+    Returns the order in which the queries should be executed using topological sort.
+    Inspired by https://gitlab.com/ericvsmith/toposort/-/blob/master/src/toposort.py
+    """
+    tmp_dep_graph = {
+        item["id"]: set(dep for dep in item["subqueries"])
+        for item in query_plan.dict()["query_graph"]
+    }
+
+    def topological_sort(
+        dep_graph: dict[int, set[int]]
+    ) -> Generator[set[int], None, None]:
+        while True:
+            ordered = set(item for item, dep in dep_graph.items() if len(dep) == 0)
+            if not ordered:
+                break
+            yield ordered
+            dep_graph = {
+                item: (dep - ordered)
+                for item, dep in dep_graph.items()
+                if item not in ordered
+            }
+        if len(dep_graph) != 0:
+            raise ValueError(
+                f"Circular dependencies exist among these items: {{{', '.join(f'{key}:{value}' for key, value in dep_graph.items())}}}"
+            )
+
+    result = []
+    for d in topological_sort(tmp_dep_graph):
+        result.extend(sorted(d))
+    return result
+
+
+def execute_queries_in_order(query_plan: QueryPlan) -> dict[int, QueryAnswer]:
+    """
+    Executes the queries in the query plan in the correct order.
+    """
+    execution_order = get_execution_order(query_plan)
+    query_dict = {q.id: q for q in query_plan.query_graph}
+    subquery_answers = {}
+    for query_id in execution_order:
+        subquery_answers[query_id] = query_dict[query_id].execute(
+            subquery_answers=[
+                subquery_answers[i] for i in query_dict[query_id].subqueries
+            ]
+        )
+    return subquery_answers
+
+
+async def aexecute_queries_in_order(query_plan: QueryPlan) -> dict[int, QueryAnswer]:
```

Comment:
what are your thoughts on implementing this as a method in queryplan?

Reply:
sure, makes way more sense, was just to tired to see that...